### PR TITLE
Refactor: Extract out repitition of start-stop node pattern

### PR DESCRIPTION
**Pull request by AryanGodara**
## Overview
I've extracted out the common pattern of
1. Initialize nodes
2. Start nodes
3. Check failures
4. Test cleanup / defer call to stop node and free resources
into separate helper functions to remove redundancy and reduce scope of common error in these.

## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

## Summary by CodeRabbit

- **Tests**
	- Enhanced test suite for node operations with streamlined startup and cleanup processes.
	- Improved error handling and maintainability across various node-related test functions.
	- Introduced `NodeType` enum to clarify node types in tests.
	- Refactored numerous test functions to use centralized initialization and cleanup logic, improving test readability.
	- Adjusted logic in test functions for more accurate simulation of node behaviors and interactions.

### CODE REVIEW

**Comment by AryanGodara** [comment]
@MSevey This is the first separate PR, only for startAndStopNode part. Sorry for the delay in opening this 😅

File: node/full_node_test.go [hunk 1]
```diff
@@ -27,27 +27,26 @@ import (
...
-	assert := assert.New(t)
+	node := initAndStartNodeWithCleanup(ctx, t, FullNodeType).(*FullNode)
+	require.NotNil(t, node)
 
-	peerID := getPeerID(assert)
-	verifyTransactions(node, peerID, assert)
-	verifyMempoolSize(node, assert)
+	peerID := getPeerID(t)
+	verifyTransactions(t, node, peerID)
+	verifyMempoolSize(t, node)
```
Comment (MSevey) [inline node/full_node_test.go:42]: looks like these changes fell in to this PR. let's remove.

File: node/full_node_test.go [hunk 2]
```diff
@@ -27,27 +27,26 @@ import (
...
 }
 
 // Tests that the node is able to sync multiple blocks even if blocks arrive out of order
 func TestTrySyncNextBlockMultiple(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
-	node, signingKey := setupTestNode(ctx, t, "full")
+	node, signingKey := setupTestNode(ctx, t, FullNodeType)
```
Comment (MSevey) [inline node/full_node_test.go:49]: looks like these changes fell in to this PR. let's remove.

File: node/full_node_test.go [hunk 3]
```diff
@@ -163,33 +157,35 @@ func setupMockApplication() *mocks.Application {
...
 // getPeerID generates a peer ID
-func getPeerID(assert *assert.Assertions) peer.ID {
-	key := generateSingleKey()
+func getPeerID(t *testing.T) peer.ID {
+	key, err := generateSingleKey()
+	assert.NoError(t, err)
+
 	peerID, err := peer.IDFromPrivateKey(key)
-	assert.NoError(err)
+	assert.NoError(t, err)
```
Comment (MSevey) [inline node/full_node_test.go:171]: looks like these changes fell in to this PR. let's remove.

File: node/full_node_test.go [hunk 4]
```diff
@@ -163,33 +157,35 @@ func setupMockApplication() *mocks.Application {
...
 	return peerID
 }
 
 // verifyTransactions checks if transactions are valid
-func verifyTransactions(node *FullNode, peerID peer.ID, assert *assert.Assertions) {
+func verifyTransactions(t *testing.T, node *FullNode, peerID peer.ID) {
```
Comment (MSevey) [inline node/full_node_test.go:176]: looks like these changes fell in to this PR. let's remove.

File: node/full_node_test.go [hunk 5]
```diff
@@ -163,33 +157,35 @@ func setupMockApplication() *mocks.Application {
...
 		})
-		assert.NoError(err)
+		assert.NoError(t, err)
 	}
 }
 
 // verifyMempoolSize checks if the mempool size is as expected
-func verifyMempoolSize(node *FullNode, assert *assert.Assertions) {
-	assert.NoError(testutils.Retry(300, 100*time.Millisecond, func() error {
+func verifyMempoolSize(t *testing.T, node *FullNode) {
```
Comment (MSevey) [inline node/full_node_test.go:187]: looks like these changes fell in to this PR. let's remove.

File: node/node_test.go [hunk 6]
```diff
@@ -18,67 +18,89 @@ import (
...
 
+// NodeType defines the type of a node using iota for an enumerated type.
+type NodeType int
+
+const (
+	// FullNode represents a full node type.
+	FullNodeType NodeType = iota
+	// LightNode represents a light node type.
+	LightNodeType
+)
```
Comment (MSevey) [inline node/node_test.go:29]: looks like these changes fell in to this PR. let's remove.

The impact of this change is lost in the diff of the `startNodeWithCleanup` changes. 
I'd rather hold of on this to be in a separate PR to better understand it's impact on these changes too
```
ln := initAndStartNodeWithCleanup(ctx, t, LightNodeType).(*LightNode)
```

**Review by MSevey — CHANGES_REQUESTED** [review]
Good start, much easier to review.
A few more requests to clean up this PR to the specific change, which also might help understand the failing CI.

**Review by MSevey — CHANGES_REQUESTED** [review]
Looking good. Some minor clean up and then I think we are good to go assuming the CI is passing.

**Comment by AryanGodara** [comment]
@MSevey This one's ready for review :D

**Review by Manav-Aggarwal — APPROVED** [review]

**Review by MSevey — APPROVED** [review]

File: node/helpers_test.go [hunk 7]
```diff
@@ -47,15 +45,12 @@ func TestGetNodeHeight(t *testing.T) {
...
-	defer func() {
-		assert.NoError(fullNode.Stop())
-	}()
 
-	assert.NoError(lightNode.Start())
-	defer func() {
-		assert.NoError(lightNode.Stop())
-	}()
+	fullNode.(*FullNode).dalc = dalc
+	fullNode.(*FullNode).blockManager.SetDALC(dalc)
```
Comment (tzdybal) [inline node/helpers_test.go:50]: Why those lines are introduced? `dalc` is already passed to `createAndConfigureNode`?

**Review by tzdybal — COMMENTED** [review]
I really like the change, but left some questions and suggestions.

File: node/full_client_test.go [hunk 8]
```diff
@@ -969,6 +919,8 @@ func TestStatus(t *testing.T) {
 	rpc.node.Store.SetHeight(ctx, latestBlock.Height())
 	require.NoError(err)
 
+	startNodeWithCleanup(t, rpc.node)
```
[the remark this answers is not in the document]
Comment (AryanGodara) [reply node/full_client_test.go:922]: I think you're right. I'll check this and fix this with the dalc code in a new PR :D

**Pull request merged**